config/v1/types_infrastructure.go

```diff
@@ -706,6 +701,55 @@ type GCPResourceTag struct {
 	Value string `json:"value"`
 }
 
+// CloudLoadBalancerConfig contains IPs for API, API-Int and Ingress Cloud Load Balancers
+// +union
+// +kubebuilder:validation:XValidation:rule="has(self.dnsType) && self.dnsType == 'ClusterHosted' ? has(self.clusterHosted) && size(self.clusterHosted.apiIntLoadBalancerIPs) > 0 : true",message="when dnsType is ClusterHosted, API-Int Load Balancer IP addresses have to specified with clusterHostedDNS"
```

Comment:
We just need to make sure that clusterHosted is not set when it's the wrong DNSType, that allows you to validate as you wanted
```suggestion
// +kubebuilder:validation:XValidation:rule="has(self.dnsType) && self.dnsType != 'ClusterHosted' ? !has(self.clusterHosted) : true",message="clusterHosted is only permitted when dnsType is ClusterHosted"
```

Reply:
I keep changing my mind about this. So, added this more involved check. But, I think it is fair to just make sure that clusterHosted is set when DNSType is `ClusterHosted`.